httpcore/_sync/http11.py

```diff
@@ -127,14 +127,14 @@ def _send_request_body(self, request: Request) -> None:
             event = h11.Data(data=chunk)
             self._send_event(event, timeout=timeout)
 
-        event = h11.EndOfMessage()
-        self._send_event(event, timeout=timeout)
+        self._send_event(h11.EndOfMessage(), timeout=timeout)
 
     def _send_event(
-        self, event: H11Event, timeout: Optional[float] = None
+        self, event: h11.Event, timeout: Optional[float] = None
     ) -> None:
         bytes_to_send = self._h11_state.send(event)
-        self._network_stream.write(bytes_to_send, timeout=timeout)
+        if bytes_to_send is not None:
```

Comment:
we know we don't send ConnectionClosed so this check is redundant

[send](https://h11.readthedocs.io/en/latest/api.html#h11.Connection.send) should be defined using [overload](https://docs.python.org/3/library/typing.html#typing.overload)

Reply:
👍 I'll see if I can contribute that upstream

Reply:
Oof, this would actually be a pretty big change upstream as the following is not a valid overload:
```python
    @overload
    def send(self, event: ConnectionClosed) -> None:
        ...

    @overload
    def send(self, event: Event) -> bytes:
        ...

    def send(self, event: Event) -> Optional[bytes]:
```

```
h11/_connection.py:505: error: Overloaded function signatures 1 and 2 overlap with incompatible return types  [misc]
```

Reply:
I'm also exploring a generic

```python
SendType = TypeVar('SendType', None, bytes)

class Event(ABC, Generic[SendType]):
    ...
```

but this would break type-checks for any downstream library that uses `Event` without a value for the generic e.g.
```
h11/_state.py:192: error: Missing type parameters for generic type "Event"  [type-arg]
```